### PR TITLE
Implement a simple IPC mechanism in preparation for the internal FSMonitor

### PR DESCRIPTION
**Pull request by dscho**
This topic branch introduces support for a super simply command server, backed by Unix sockets (and named pipes on Windows). All it does is to listen for short, textual commands, and respond with a reply, then close the connection.

The intention is to support an upcoming implementation of a built-in, Git-aware FSMonitor.

### CODE REVIEW

**Comment by dscho** [comment]
@jeffhostetler do you want to work on the task where simple IPC learns to potentially service multiple clients in parallel? If so, do you think it should be an add-on, or should it be squashed into this PR?

**Comment by derrickstolee** [comment]
> @jeffhostetler do you want to work on the task where simple IPC learns to potentially service multiple clients in parallel? If so, do you think it should be an add-on, or should it be squashed into this PR?

Unless there is a good reason why something done here is blocking that feature, I think it would be good to get this reviewed and merged before he updates the protocol with that part.

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c]: I was surprised by this syntax, but then I took a look:
```
$ git grep "while (1)" | wc -l
114
$ git grep "for (;;)" | wc -l
149
```

File: t/t0052-simple-ipc.sh
Comment (derrickstolee) [inline t/t0052-simple-ipc.sh:10]: Should this be a prereq defined in `test-lib.sh`? That way we can require SUPPORTS_SIMPLE_IPC when we eventually make our fsmonitor tests.

File: t/t0052-simple-ipc.sh
Comment (derrickstolee) [inline t/t0052-simple-ipc.sh:22]: :+1:

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c:611]: nit: this commit could maybe use more text in the message body. Perhaps:

```
We previously implemented simple Inter-Process Communication (IPC) on
Windows using named pipes. Create an equivalent implementation for
Unix-based platforms using sockets.

(insert any subtleties about how this differs from the Windows implementation)

The tests in t0052-simple-ipc.sh will automatically start running on Unix
platforms with this change due to the response from "test-tool simple-ipc
SUPPORTS_SIMPLE_IPC" changing automatically with the change here in
simple-ipc.h.
```

File: simple-ipc.h
Comment (derrickstolee) [inline simple-ipc.h:24]: I love how simple this update is.

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c]: Do we want this path to be named specifically with `fsmonitor`, or do we want something more generic?

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c:693]: this trace2 region is missing from the Windows implementation. Can we add it there? It would be nice to see what we can do to make these as equivalent as possible?

Perhaps we need to break this method into sub-parts that have platform-specific implementations, but the general structure is platform agnostic?

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c]: this differs from the `for (;;) {` you used before. Can we keep consistent?

Comment (derrickstolee) [reply simple-ipc.c]: But that was in the `ipc_send_command()` method. Should we have `while (server->active)` here?

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c]: Now that we are implementing these features simultaneously in both platforms, I wonder if we should be changing the file structure a bit. It is hard to tell from the patch context that we are in the `#ifdef WINDOWS` region.

Perhaps we create an `ipc` directory, and have a split of files this way:

```
ipc/
    simple-ipc.h
    simple-ipc-internal.h (stores struct and method prototypes used by both *-ipc.c files)
    simple-ipc.c  (stores platform-agnostic logic)
    windows-ipc.c
    unix-ipc.c
```

For example, the different platform-specific pieces of `ipc_listen_for_commands()` could be declared in `simple-ipc-internal.h` and implemented in `<platform>-ipc.c`, but `ipc_listen_for_commands()` itself is declared in `simple-ipc.h` and implemented in `simple-ipc.c`.

Then, the `Makefile` could decide which of `windows-ipc.c` or `unix-ipc.c` to compile and link.

I know it would be a lot of work to rearrange this, but do you think this is a better direction for long term maintainability?

File: t/helper/test-simple-ipc.c
Comment (derrickstolee) [inline t/helper/test-simple-ipc.c]: Is there a way to generalize this retry loop so we don't need to reimplement it each time?

File: simple-ipc.c
Comment (derrickstolee) [inline simple-ipc.c]: Is this only for debugging purposes? Or is there a long-term benefit to having these traces?

File: pkt-line.c
Comment (derrickstolee) [inline pkt-line.c:258]: I have learned that we frequently need to make such an `int <option>` into `int flags` at a later date. Do we want to get ahead of that now, or only do the `flush_at_end -> flags` conversion if/when that happens?

File: pkt-line.c
Comment (derrickstolee) [inline pkt-line.c:429]: nit: `options` -> `flags`? I'm not set on this, so feel free to ignore.

**Review by derrickstolee — CHANGES_REQUESTED** [review]
I recommend some pretty significant shifts in the IPC code organization, so please give them some thought. Most everything else is a nit.

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: Yep. I am surprised, quite honestly, that there are any `while (1)` at all, given the strong preference by the Git maintainer ;-)

File: t/t0052-simple-ipc.sh
Comment (dscho) [reply t/t0052-simple-ipc.sh:10]: > Should this be a prereq defined in `test-lib.sh`?

I originally had this as a prereq, but since _all_ of the tests would need that prereq, I opted for the `skip_all` instead.

> That way we can require SUPPORTS_SIMPLE_IPC when we eventually make our fsmonitor tests.

That's not actually necessary, as there is _another_ prereq for the FSMonitor: not all platforms that support simple IPC support FSMonitor. Besides, the fact that FSMonitor uses simple IPC is an implementation detail that should not bleed into `t/t7527-internal-fsmonitor.sh`.

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c:611]: How about this:

    simple-ipc: implement a backend based on Unix-sockets

    We previously implemented simple Inter-Process Communication (IPC) on
    Windows using named pipes. Create an equivalent implementation for
    Unix-based platforms using Unix sockets.

    Note that there are a variety of differences between Win32 Named Pipes
    and Unix sockets, for example:

    - a Win32 Named Pipe vanishes once the listener handle is closed. Unix
      sockets do not just disappear, even if the process that created them
      exited.

    - Win32 Named Pipes exist in a name space parallel to the actual file
      system. In contrast, Unix sockets live in the same name space, i.e. a
      Unix socket cannot have the same path as an existing file and vice
      versa.

    - Win32 Named Pipes have two communication modes: "message" and "byte".
      The former allows senders to define the size of messages to be sent;
      the receiving end will not receive partial messages. Unix sockets only
      operate in the mode equivalent to "byte".

    - When a listener accepts a client via a Win32 Named Pipe, subsequent
      communication will use the same handle to the same pipe, i.e. it can
      become a bit tricky to keep the communication to several concurrent
      clients apart from one another. When a listener on a Unix socket
      accepts a new client, it gets handed a new file descriptor specific to
      that client.

    The tests in `t0052-simple-ipc.sh` will automatically start running on
    Unix platforms with this change due to the response from `test-tool
    simple-ipc SUPPORTS_SIMPLE_IPC` changing automatically with the change
    here in `simple-ipc.h`.

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: More generic!

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c:693]: I like the idea of making this as platform-agnostic as possible, but when I look through the code, it does not seem to lend itself to being abstracted, unless we want to pay a penalty (which I am not really prepared).

About that Trace2 region: absolutely 👍 . I also fail to miss the fact that the `_leave()` call is missing.

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: I made it so, even if that `active` bit was not used there (no idea why I did things differently in the Unix sockets part...).

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: I don't know that this would make it easier to read. I like to have the IPC things in one file, and it strikes me as a similar situation to `run-command.c`, where large parts are `#ifdef GIT_WINDOWS_NATIVE ... #else ... #if`. It's just that C preprocessor idioms cause this kind of readability issues, but that's not something I can change in this PR.

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: Debugging. Will remove it.

File: pkt-line.c
Comment (dscho) [reply pkt-line.c:429]: Oh, you're absolutely right!

Comment (dscho) [reply pkt-line.c:429]: ... except that we actually hand that value to `packet_read()`, which _already_ takes an `options` parameter... So I'll leave this consistent with the rest of `pkt-line.c`.

File: pkt-line.c
Comment (dscho) [reply pkt-line.c:258]: I'd rather wait until (or _ifff_) we need it...

File: simple-ipc.c
Comment (jeffhostetler) [reply simple-ipc.c:611]: WRT:

```
    - When a listener accepts a client via a Win32 Named Pipe, subsequent
       communication will use the same handle to the same pipe, i.e. it can
       become a bit tricky to keep the communication to several concurrent
       clients apart from one another. When a listener on a Unix socket
       accepts a new client, it gets handed a new file descriptor specific to
       that client.
```

Windows Named Pipes actually behave a little differently than that.
The server creates the initial server-pipe (which reserves the pathname
in the Named Pipe File System) and allows it to receive the first client
connection.  That part is true.  But when either side closes that connection,
the pipe is dead.  The pathname is removed from the NPFS when the other
side closes their handle.

For the server to respond to multiple clients (either in series or currently),
it needs to open more than one server pipe handles on the same pathname.
(The server should set the ACL on the initial server-pipe handle, because the
ACL can't be changed once there are multiple handles.)

So yes, the server needs to juggle a list/queue of open server handles.  The next
incoming client connection will connect (randomly to) one of the open
server handles.  The server needs to pull that from the "ready list/queue" and
dispatch it to a worker thread for the duration of the conversation/protocol.
The server needs to open a new server handle on the existing pathname
to replace the consumed one in the queue.

Most of the example code on the net show a server opening a single server
handle, waiting for a connection, dispatching it and then opening a new
single server handle to replace it (and before the worker thread can close
the connected handle).  There are 2 races in that model: (1) the worker hangs
up too quickly and (2) the named pipe is non-responsive during that the interval
(after the connection is accepted and until the new server handle is opened).

(Creating a named-pipe client handle does secretly have a timeout in it, but
it is still possible for the client to timeout and fail to get a client handle (when
you have a high arrival rate (ask me how I know...)))

By letting the server pre-create multiple server handles (mostly) prevents
clients from starving, and implicitly prevents the other race.  Conceptually,
this server handle list/queue acts like the listen(n) in the socket code.

**Comment by derrickstolee** [comment]
@dscho do you plan to rebase this topic to squash in the `fixup!` commits? or are you waiting on some more feedback?

**Review by jeffhostetler — CHANGES_REQUESTED** [review]
I'd like to put a hold on this for a while.  Reasons forthcoming.

File: convert.c
Comment (jeffhostetler) [inline convert.c:902]: There's an odd asymmetry about this.  We only have 1 caller of `_from_buf()` and `_from_fd()`, yet only one gets the 3rd argument.

I'm wondering if we should update both to have consistent behavior -- OR -- move the flush to here following the call (when !err).

File: simple-ipc.c
Comment (jeffhostetler) [inline simple-ipc.c]: The commit message for this commit (https://github.com/microsoft/git/pull/277/commits/031698d3d5a55ab9e993ad4c7915040c08a44ac2)
says it doesn't use pkt-line routines.

File: convert.c
Comment (dscho) [reply convert.c:902]: I did not want to introduce unneeded code changes. Currently, I do not need the `_from_fd()` function, and it is not even clear whether there will ever be a use case where `_from_fd()` needs to _not_ send a flush packet at the end.

Does that make sense?

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c]: Whoops.

Comment (dscho) [reply simple-ipc.c]: I proposed a new commit message in https://github.com/microsoft/git/pull/277/commits/4e01e87166ea0487fb494dfa1302284016f8b75d (as a `squash!` because [we do not have a `reword!` yet](https://github.com/gitgitgadget/git/issues/259)):

```
    Add support for very simple Inter-Process Communication

    For the upcoming `fsmonitor--daemon` command, we need some way to send a
    command from some Git process to that daemon, and to receive an answer.
    In other words, we need to support Inter-Process Communication.

    By nature, Inter-Process Communication requires OS-specific
    implementations.

    Instead of using full-blown TCP/UDP socket communication (which comes
    with all kinds of challenges such as: restricting access permissions,
    collisions when multiple instances might want to listen on the same
    port, etc), we use named pipes on Windows, and plan on implementing a
    backend based on Unix sockets for Linux/Unix later.

    For now, we use a super-simple protocol: identified by a path into
    .git/, the daemon listens for clients, reads one command, then sends the
    answer and closes the connection.

    To avoid redundant coding patterns, we use the `pkt-line` protocol that
    we just libified to be able to use here: we really, really do not want
    our simple IPC client to take down the entire Git process with it when
    it could not talk to `fsmonitor--daemon`, but would rather instead fall
    back to a full scan of all tracked files.
```

(I adjusted the second-to-last paragraph minimally, and the last paragraph rather a lot.)

File: simple-ipc.c
Comment (dscho) [reply simple-ipc.c:611]: @jeffhostetler can I ask you to implement this? You are the domain expert about this, and I don't trust myself to get it right...

**Comment by dscho** [comment]
> @dscho do you plan to rebase this topic to squash in the `fixup!` commits? or are you waiting on some more feedback?

I'm just majorly side-tracked with https://github.com/git-for-windows/git/issues/2674, is all ;-)

File: convert.c
Comment (jeffhostetler) [reply convert.c:902]: that's fine.  just asking.

File: simple-ipc.c
Comment (jeffhostetler) [reply simple-ipc.c:611]: Let me take a look at restructuring things and see what I can come up with.  I think it would be easier to slip this in at the IPC layer in isolation before worrying about mingling it with FSMonitor details.

Comment (dscho) [reply simple-ipc.c:611]: Okay.

**Comment by jeffhostetler** [comment]
I'd like to close this PR in favor of #281

**Pull request closed without merging**
